python/pyspark/sql/types.py

```diff
@@ -181,6 +182,29 @@ class StringType(AtomicType, metaclass=DataTypeSingleton):
     pass
 
 
+class VarcharType(AtomicType):
```

Comment:
`metaclass=DataTypeSingleton`?

Reply:
I firstly tried this, but it cause initialization error due to the `__call__` in `DataTypeSingleton`:
a type mixin with `DataTypeSingleton` should support constructor without parameters.

Reply:
https://issues.apache.org/jira/browse/SPARK-39794